plugins/node/opentelemetry-instrumentation-bunyan/src/OpenTelemetryBunyanStream.ts

```diff
@@ -0,0 +1,162 @@
+  [ERROR]: SeverityNumber.ERROR,
+  [FATAL]: SeverityNumber.FATAL,
+};
+
+const EXTRA_SEV_NUMS = [
+  SeverityNumber.TRACE2,
+  SeverityNumber.TRACE3,
+  SeverityNumber.TRACE4,
+  SeverityNumber.DEBUG2,
+  SeverityNumber.DEBUG3,
+  SeverityNumber.DEBUG4,
+  SeverityNumber.INFO2,
+  SeverityNumber.INFO3,
+  SeverityNumber.INFO4,
+  SeverityNumber.WARN2,
+  SeverityNumber.WARN3,
+  SeverityNumber.WARN4,
+  SeverityNumber.ERROR2,
+  SeverityNumber.ERROR3,
+  SeverityNumber.ERROR4,
+  SeverityNumber.FATAL2,
+  SeverityNumber.FATAL3,
+  SeverityNumber.FATAL4,
+];
+
+function severityNumberFromBunyanLevel(lvl: number) {
+  // Fast common case: one of the known levels
+  const sev = OTEL_SEV_NUM_FROM_BUNYAN_LEVEL[lvl];
+  if (sev !== undefined) {
+    return sev;
+  }
+
+  // Otherwise, scale the Bunyan level range -- 10 (TRACE) to 70 (FATAL+10)
+  // -- onto the extra OTel severity numbers (TRACE2, TRACE3, ..., FATAL4).
+  // Values below bunyan.TRACE map to SeverityNumber.TRACE2, which may be
+  // considered a bit weird, but it means the unnumbered levels are always
+  // just for exactly values.
+  const relativeLevelWeight = (lvl - 10) / (70 - 10);
+  const otelSevIdx = Math.floor(relativeLevelWeight * EXTRA_SEV_NUMS.length);
+  const cappedOTelIdx = Math.min(
+    EXTRA_SEV_NUMS.length - 1,
+    Math.max(0, otelSevIdx)
+  );
+  const otelSevValue = EXTRA_SEV_NUMS[cappedOTelIdx];
+  return otelSevValue;
+}
+
+/**
+ * A Bunyan stream for sending log records to the OpenTelemetry Logs Bridge API.
+ */
+export class OpenTelemetryBunyanStream {
+  private _otelLogger: Logger;
+
+  constructor() {
+    this._otelLogger = logs.getLogger(
+      DEFAULT_INSTRUMENTATION_SCOPE_NAME,
+      DEFAULT_INSTRUMENTATION_SCOPE_VERSION
+    );
+  }
+
+  /**
+   * Convert from https://github.com/trentm/node-bunyan#log-record-fields
+   * to https://opentelemetry.io/docs/specs/otel/logs/data-model/
+   *
+   * Dev Notes:
+   * - We drop the Bunyan 'v' field. It is meant to indicate the format
+   *   of the Bunyan log record. FWIW, it has always been `0`.
+   * - The standard Bunyan `hostname` and `pid` fields are removed because they
+   *   are redundant with the OpenTelemetry `host.name` and `process.pid`
+   *   Resource attributes, respectively. This code cannot change the
+   *   LoggerProvider's `resource`, so getting the OpenTelemetry equivalents
+   *   depends on the user using relevant OpenTelemetry resource detectors.
+   *   "examples/telemetry.js" shows using HostDetector and ProcessDetector for
+   *   this.
+   * - The Bunyan `name` field *could* naturally map to OpenTelemetry's
+   *   `service.name` resource attribute. However, that is debatable, as some
+   *   users might use `name` more like a log4j logger name.
+   * - Strip the `trace_id` et al fields that may have been added by the
+   *   the _emit wrapper.
+   */
+  write(rec: Record<string, any>) {
+    // eslint-disable-next-line @typescript-eslint/no-unused-vars
+    const {
```

Comment:
Do we want to control the log level that will generate log telemetry here? Python supports a env variable called OTEL_PYTHON_LOG_LEVEL more info [here](https://opentelemetry-python-contrib.readthedocs.io/en/latest/instrumentation/logging/logging.html#envvar-OTEL_PYTHON_LOG_LEVEL), is a common ask in Azure side to control which level of logs to get telemetry for

Reply:
I think this would be fairly straightforward. A starter patch, without error-checking:

```diff
diff --git a/plugins/node/opentelemetry-instrumentation-bunyan/src/instrumentation.ts b/plugins/node/opentelemetry-instrumentation-bunyan/src/instrumentation.ts
index 8fd0d5bc..0007a119 100644
--- a/plugins/node/opentelemetry-instrumentation-bunyan/src/instrumentation.ts
+++ b/plugins/node/opentelemetry-instrumentation-bunyan/src/instrumentation.ts
@@ -160,7 +160,7 @@ export class BunyanInstrumentation extends InstrumentationBase<
     logger.addStream({
       type: 'raw',
       stream: new OpenTelemetryBunyanStream(),
-      level: logger.level(),
+      level: process.env.OTEL_NODE_LOG_LEVEL || logger.level(), // XXX obviously missing error checking
     });
   }
```

To be clear, my understanding is that this level would be independent of the level used for the other output streams on the logger. So instrumenting the following with `OTEL_NODE_LOG_LEVEL=trace`:

```js
const logger = bunyan.createLogger({name: 'myapp', level: 'debug'})
logger.debug('hi at debug level')
logger.trace('hi at trace level')
```

would result in just "hi at debug level" being written to stdout, but both log messages being sent to the Logs Bridge API.

- Are the supported level values the *OTel log severity names* (inferred from the table for SeverityNumber here https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-severitytext) or are they the Bunyan log level names. Conveniently these are a 1:1 mapping for *Bunyan*, but not for Winston (https://github.com/winstonjs/winston#logging-levels) for example.
- What does Java-land do, if anything?
- Should this be configurable on `new BunyanInstrumentation({...})` as well? Probably.
- Is there an prior art for "NODE" or "JS" or "NODEJS" in `OTEL_*` environment variable names?

One hesitation I have here: I know Bunyan can support having different levels on different output streams/appenders/destinations. I'm not positive for Pino and Winston. What do you think about defering supporting this configuration until I (or someone else) has worked on equivalent instrumentation for those logging frameworks?

Reply:
Yeah my comment was more about adding the option to configure it, not necessary by env variable, I don't see any OTEL_NODE or similar env variable being used anywhere, maybe we can just allow it by Instrumentation config but these need to be applicable to all Instrumentations that will create logs so we need to use OTel severity names to configure it. We can follow up later and should be fine for me as well, we just need to support it eventually, if you don't add it most likely I will add it very soon. This is becoming a higher priority for us in Azure side :)

Reply:
Ah, okay. I think it would be good and useful to add the configuration option to the constructor for this.

> maybe we can just allow it by Instrumentation config 

+1

> but these need to be applicable to all Instrumentations that will create logs so we need to use OTel severity names to configure it.

Hrm, I'm still undecided/unsure on this. If I am configuring the log level for my Winston logger, for example, should I not be able to specify the "crit" level defined at https://github.com/winstonjs/winston#logging-levels?
The Logs data model somewhat goes out of its way to not define specific level/severity names: https://opentelemetry.io/docs/specs/otel/logs/data-model/#field-severitytext

Or perhaps we need to clearly show the mappings of Bunyan, Pino, Winston log levels in this table: https://opentelemetry.io/docs/specs/otel/logs/data-model-appendix/#appendix-b-severitynumber-example-mappings
and then the config var accepts those "SeverityNumber" *strings* (and/or perhaps the SeverityNumber integer values).

Reply:
I've spun this out to a separate issue https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1821
after some discussion on yesterday's OTel JS SIG.